File: solutions/cpp/high-scores/1/high_scores.cpp
#include "high_scores.h"

#include <algorithm>
#include <iostream>

namespace arcade {

std::vector<int> HighScores::list_scores() {
    return scores;
}

int HighScores::latest_score() {
    return scores.back();
}

int HighScores::personal_best() {
    int highest{0};

    if (!scores.empty()) {
        highest = scores[0];
    }

    for (int score : scores) {
        if (score > highest) {
            highest = score;
        }
    }
    return highest;
}

std::vector<int> HighScores::top_three() {
    std::vector<int> cache{scores};
    std::vector<int> result;

    std::sort(cache.begin(), cache.end(), [](int a, int b) {
       return a > b;
    });

    size_t i{0};
    while (i < cache.size() && i < 3) {
        result.push_back(cache[i]);
        ++i;
    }

    return result;
}

}  // namespace arcade

File: solutions/go/high-scores/1/high_scores.go
package highscores

import "sort"

type HighScores struct {
	scores []int
}

// NewHighScores returns a new HighScores object.
func NewHighScores(scores []int) *HighScores {
	s := &HighScores{
		scores: scores,
	}
	return s
}

// Scores returns all the scores.
func (s *HighScores) Scores() []int {
	return s.scores
}

// Latest returns the latest (last) score.
func (s *HighScores) Latest() int {
	return s.scores[len(s.scores)-1]
}

// PersonalBest returns the best (highest) score.
func (s *HighScores) PersonalBest() int {
	if len(s.scores) == 0 {
		return 0
	} else if len(s.scores) == 1 {
		return s.scores[0]
	}

	// establish current for future compares
	highest := s.scores[0]
	for _, v := range s.scores {
		highest = max(v, highest)
	}
	return highest
}

// TopThree returns the top three scores.
func (s *HighScores) TopThree() []int {
	c := make([]int, len(s.scores))
	copy(c, s.scores)
	sort.Slice(c, func(i, j int) bool {
		return c[i] > c[j]
	})
	// guard against less than three in slice
	if len(c) < 3 {
		return c
	}
	return c[:3]
}
